src/main/java/fi/aalto/cs/apluscourses/intellij/model/APlusProject.java

```diff
@@ -0,0 +1,125 @@
+package fi.aalto.cs.apluscourses.intellij.model;
+
+import com.intellij.openapi.module.ModuleManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.roots.libraries.LibraryTable;
+import com.intellij.openapi.roots.libraries.LibraryTablesRegistrar;
+import com.intellij.util.messages.MessageBus;
+import fi.aalto.cs.apluscourses.model.Component;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Objects;
+import org.jetbrains.annotations.NotNull;
+
+public class APlusProject {
+  private final Project project;
+
+  public APlusProject(Project project) {
+    this.project = project;
+  }
+
+  @NotNull
+  public Path getBasePath() {
+    return Paths.get(Objects.requireNonNull(project.getBasePath()));
+  }
+
+  @NotNull
+  public ModuleManager getModuleManager() {
+    return ModuleManager.getInstance(project);
+  }
+
+  @NotNull
+  public MessageBus getMessageBus() {
+    return project.getMessageBus();
+  }
+
+  @NotNull
+  public LibraryTable getLibraryTable() {
+    return LibraryTablesRegistrar.getInstance().getLibraryTable(project);
+  }
+
+  @NotNull
+  public Path getLibraryPath(String name) {
+    return  Paths.get("lib", name);
+  }
+
+  @NotNull
+  public Path getModulePath(String name) {
+    return Paths.get(name);
+  }
+
+  public int resolveLibraryState(String name) {
+    return new LibraryStateResolver(name).resolveState();
+  }
+
+  public int resolveModuleState(String name) {
+    return new ModuleStateResolver(name).resolveState();
+  }
+
+  private abstract class ComponentStateResolver {
```

Comment:
In this case this abstraction adds quite a bit of complexity as opposed to just having the state resolving code in `resolveLibraryState` and `resolveModuleState` with some minor code duplication. Could be useful though if we add different types of components later

Reply:
Yeah, now that I look this, it's terribly complex. As far as I can see, there ain't gonna be other components than libraries and modules, so following YAGNI principle, I'll adopt your suggestion here.